diff --git a/code/lib/list.h b/code/lib/list.h
index 3ce9553..6ab0013 100644
--- a/code/lib/list.h
+++ b/code/lib/list.h
@@ -100,6 +100,8 @@ class SortedList : public List<T> {
     void SanityCheck() const;  // has this list been corrupted?
     void SelfTest(T *p, int numEntries);
     // verify module is working
+    // in a sorted list
+    void Append(T item) { Insert(item); }
 
    private:
     int (*compare)(T x, T y);  // function for sorting list elements
@@ -108,7 +110,8 @@ class SortedList : public List<T> {
         Insert(item);
     }                                      // *pre*pending has no meaning
                                            //	in a sorted list
-    void Append(T item) { Insert(item); }  // neither does *ap*pend
+   // void Append(T item) { Insert(item); }  // neither does *ap*pend
+   
 };
 
 // The following class can be used to step through a list.
diff --git a/code/machine/timer.cc b/code/machine/timer.cc
index c18c317..a0736c3 100644
--- a/code/machine/timer.cc
+++ b/code/machine/timer.cc
@@ -64,7 +64,7 @@ void Timer::CallBack() {
 //----------------------------------------------------------------------
 
 void Timer::SetInterrupt() {
-    if (!disable) {
+    if (!disable && false) {
         int delay = TimerTicks;
 
         if (randomize) {
diff --git a/code/test/add.c b/code/test/add.c
index e80a2df..2d016cf 100644
--- a/code/test/add.c
+++ b/code/test/add.c
@@ -15,6 +15,7 @@ int main() {
     for(i=0; i<10; i++){
 	 Write("C\n", 2,1);
      }
-    Halt();
+    Exit(0);
+    //Halt();
     /* not reached */
 }
diff --git a/code/test/exec.c b/code/test/exec.c
index 7e3b46a..717e866 100644
--- a/code/test/exec.c
+++ b/code/test/exec.c
@@ -20,5 +20,6 @@ int main() {
     } else{
         Write("Exec failed\n",12,stdout);
     }
-    Halt();
+    Exit(0);
+    //Halt();
 }
diff --git a/code/threads/pcb.cc b/code/threads/pcb.cc
index a740854..065e717 100644
--- a/code/threads/pcb.cc
+++ b/code/threads/pcb.cc
@@ -70,6 +70,7 @@ int PCB::Exec(char* filename, int id) {
     // nên khi hàm này kết thúc thì giá trị của biến này cũng bị xóa
     // Đừng hỏi tôi đã mất bao lâu để nhận ra điều này :)
     this->thread->Fork(StartProcess_2, &this->thread->processID);
+    kernel->currentThread->Yield();
 
     multex->V();
     // Trả về id.
diff --git a/code/threads/scheduler.cc b/code/threads/scheduler.cc
index 1d52916..ed111c4 100644
--- a/code/threads/scheduler.cc
+++ b/code/threads/scheduler.cc
@@ -22,7 +22,14 @@
 #include "debug.h"
 #include "scheduler.h"
 #include "main.h"
-
+int compare(Thread* x, Thread* y) {
+    if (x->priority > y->priority)
+        return 1;
+    else if (x->priority < y->priority)
+        return -1;
+    else
+        return 0;
+}
 //----------------------------------------------------------------------
 // Scheduler::Scheduler
 // 	Initialize the list of ready but not running threads.
@@ -30,7 +37,8 @@
 //----------------------------------------------------------------------
 
 Scheduler::Scheduler() {
-    readyList = new List<Thread *>;
+    //readyList = new List<Thread *>;
+    readyList = new SortedList<Thread *>(&compare);
     toBeDestroyed = NULL;
 }
 
@@ -54,7 +62,7 @@ void Scheduler::ReadyToRun(Thread *thread) {
     DEBUG(dbgThread, "Putting thread on ready list: " << thread->getName());
 
     thread->setStatus(READY);
-    readyList->Append(thread);
+    readyList->Append(thread);	
 }
 
 //----------------------------------------------------------------------
@@ -164,3 +172,4 @@ void Scheduler::Print() {
     cout << "Ready list contents:\n";
     readyList->Apply(ThreadPrint);
 }
+
diff --git a/code/threads/scheduler.h b/code/threads/scheduler.h
index a96f228..7151bd2 100644
--- a/code/threads/scheduler.h
+++ b/code/threads/scheduler.h
@@ -35,8 +35,9 @@ class Scheduler {
     // SelfTest for scheduler is implemented in class Thread
 
    private:
-    List<Thread*>* readyList;  // queue of threads that are ready to run,
-                               // but not running
+    //List<Thread*>* readyList;  // queue of threads that are ready to run,
+                               // but 
+    SortedList<Thread*>* readyList;
     Thread* toBeDestroyed;     // finishing thread to be destroyed
                                // by the next thread that runs
 };
diff --git a/code/threads/thread.cc b/code/threads/thread.cc
index be9f7d6..fadcb38 100644
--- a/code/threads/thread.cc
+++ b/code/threads/thread.cc
@@ -21,6 +21,7 @@
 #include "switch.h"
 #include "synch.h"
 #include "sysdep.h"
+#include <stdlib.h>
 
 // this is put at the top of the execution stack, for detecting stack overflows
 const int STACK_FENCEPOST = 0xdedbeef;
@@ -45,6 +46,8 @@ Thread::Thread(char *threadName, bool _has_dynamic_name /*=false*/) {
                                  // of machine registers
     }
     space = NULL;
+    priority = rand()%10;
+    printf("priority: %d, name: %s \n", priority, name);
 }
 
 //----------------------------------------------------------------------
@@ -100,7 +103,10 @@ void Thread::Fork(VoidFunctionPtr func, void *arg) {
 
     oldLevel = interrupt->SetLevel(IntOff);
     scheduler->ReadyToRun(this);  // ReadyToRun assumes that interrupts
-                                  // are disabled!
+                                  // are disab
+    if (this->priority > kernel->currentThread->priority){
+	    kernel->currentThread->Yield();
+    }
     (void)interrupt->SetLevel(oldLevel);
 }
 
diff --git a/code/threads/thread.h b/code/threads/thread.h
index 3fbf156..82a27f1 100644
--- a/code/threads/thread.h
+++ b/code/threads/thread.h
@@ -89,6 +89,7 @@ class Thread {
 
     int processID;
     int parrentID;
+    int priority;
     int exitStatus;
     void FreeSpace() {
         if (space != 0) delete space;
diff --git a/code/userprog/.syscall.h.swp b/code/userprog/.syscall.h.swp
index c7ae06b..a6840e3 100644
Binary files a/code/userprog/.syscall.h.swp and b/code/userprog/.syscall.h.swp differ
diff --git a/pandc.diff b/pandc.diff
new file mode 100644
index 0000000..e69de29
